networking/nat/src/main/java/tech/pegasys/teku/networking/nat/NatManager.java

```diff
@@ -0,0 +1,423 @@
+          }
+
+          for (SafeFuture<RemoteService> future : recognizedServices.values()) {
+            future.cancel(true);
+          }
+
+          upnpService.getRegistry().removeListener(registryListener);
+          upnpService.shutdown();
+        });
+  }
+
+  private synchronized SafeFuture<RemoteService> discoverService(final String serviceType) {
+    return recognizedServices.get(serviceType);
+  }
+
+  private void initiateExternalIpQuery() {
+    discoverService(SERVICE_TYPE_WAN_IP_CONNECTION)
+        .thenAccept(
+            service -> {
+
+              // our query, which will be handled asynchronously by the jupnp library
+              GetExternalIP callback =
+                  new GetExternalIP(service) {
+
+                    /**
+                     * Override the success(ActionInvocation) version of success so that we can take
+                     * a peek at the network interface that we discovered this on.
+                     *
+                     * <p>Because the underlying jupnp library omits generics info in this method
+                     * signature, we must too when we override it.
+                     */
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void success(final ActionInvocation invocation) {
+                      RemoteService service = (RemoteService) invocation.getAction().getService();
+                      RemoteDevice device = service.getDevice();
+                      RemoteDeviceIdentity identity = device.getIdentity();
+
+                      discoveredOnLocalAddress =
+                          Optional.of(identity.getDiscoveredOnLocalAddress().getHostAddress());
+
+                      super.success(invocation);
+                    }
+
+                    @Override
+                    protected void success(final String result) {
+
+                      LOG.info(
+                          "External IP address {} detected for internal address {}",
+                          result,
+                          discoveredOnLocalAddress.get());
+
+                      externalIpQueryFuture.complete(result);
+                    }
+
+                    /**
+                     * Because the underlying jupnp library omits generics info in this method
+                     * signature, we must too when we override it.
+                     */
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void failure(
+                        final ActionInvocation invocation,
+                        final UpnpResponse operation,
+                        final String msg) {
+                      externalIpQueryFuture.completeExceptionally(new Exception(msg));
+                    }
+                  };
+              FutureUtil.ignoreFuture(upnpService.getControlPoint().execute(callback));
+            })
+        .reportExceptions();
```

Comment:
We should explicit handle exceptions and report them via the logger with a meaningful message (and probably at WARN level) instead of using reportExceptions.  Currently get a very scary `PLEASE FIX OR REPORT` error if port forwarding fails.  eg:

```
2021-01-27 11:10:11.202+10:00 | jupnp-14 | ERROR | teku-status-log | PLEASE FIX OR REPORT | Unexpected exception thrown for jupnp-14
java.util.concurrent.CompletionException: java.lang.Exception: Error: ConflictInMappingEntry (HTTP response was: 500 Internal Server Error)
        at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1093) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) [?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152) [?:?]
        at tech.pegasys.teku.networking.nat.NatManager$4.failure(NatManager.java:386) [teku-networking-nat-develop.jar:21.1.0+25-g97032cd]
        at org.jupnp.controlpoint.ActionCallback.failure(ActionCallback.java:178) [org.jupnp-2.5.2.jar:?]
        at org.jupnp.controlpoint.ActionCallback.run(ActionCallback.java:158) [org.jupnp-2.5.2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.Exception: Error: ConflictInMappingEntry (HTTP response was: 500 Internal Server Error)
        ... 8 more
```

Also, I'm getting those errors on the second run of Teku quite consistently unless I pick a different p2p port - doesn't look like the port forwards are being closed.